### PR TITLE
Wip posix ordered

### PR DESCRIPTION
**Pull request by mattbenjamin**
## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [x] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [x] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Comment by mattbenjamin** [comment]
still being fitted together--this is the freestanding implementation and it's unittests

File: .gitmodules [hunk 1]
```diff
@@ -72,3 +72,9 @@
 [submodule "src/jaegertracing/opentelemetry-cpp"]
 	path = src/jaegertracing/opentelemetry-cpp
 	url = https://github.com/open-telemetry/opentelemetry-cpp.git
+[submodule "src/lmdb-safe"]
+	path = src/lmdb-safe
+	url = https://github.com/ahupowerdns/lmdb-safe.git
+[submodule "src/rgw/driver/posix/lmdb-safe"]
```
Comment (dang) [inline .gitmodules:78]: Do we need to include the submodule twice?

Comment (mattbenjamin) [reply .gitmodules:78]: is it?  will fix

**Comment by mattbenjamin** [comment]
still need to run posixdriver interactively;  this passes the listing cache unittests, however

**Comment by mattbenjamin** [comment]
@dang it lists bucket entries, my dude:

[mbenjamin@lemon python]$ s3cmd ls s3://
2023-03-16 18:44  s3://buck1
2023-03-16 18:44  s3://buck2
2023-03-16 18:44  s3://buck3
[mbenjamin@lemon python]$ s3cmd ls s3://buck1/
1970-01-01 00:00        98304  s3://buck1/foop1

**Comment by mattbenjamin** [comment]
(forgot to fix timestamp serialization, it would appear ;)

**Comment by mattbenjamin** [comment]
sorting and inotify are working:
[mbenjamin@lemon python]$ ls -U -1 /home/mbenjamin/rgw_posix_driver/buck1
foop1
nono1
ahyaho1
[mbenjamin@lemon python]$ s3cmd ls s3://buck1
1970-01-01 00:00            0  s3://buck1/ahyaho1
1970-01-01 00:00        98304  s3://buck1/foop1
1970-01-01 00:00            0  s3://buck1/nono1

**Comment by mattbenjamin** [comment]
[mbenjamin@lemon python]$ s3cmd ls s3://buck1
2023-05-25 17:20            0  s3://buck1/adilger
2023-05-25 16:49            0  s3://buck1/ahyaho1
2023-03-16 19:03        98304  s3://buck1/foop1
2023-05-25 16:48            0  s3://buck1/nono1

**Review by dang — APPROVED** [review]

File: src/test/rgw/test_posix_bucket_cache.cc [hunk 2]
```diff
@@ -0,0 +1,435 @@
...
+{
+  MockSalBucket sb{tdir1};
+  rgw::sal::Bucket::ListParams params{};
+  params.marker.name = bucket1_marker;
+  (void) bucket_cache->list_bucket(dpp, &sb, params, 9999999 /* XXXX */, func);
+}
+
+TEST(BucketCache, ListTDir2)
+{
+  MockSalBucket sb{tdir1};
```
[the remark this answers is not in the document]
Comment (mattbenjamin) [reply src/test/rgw/test_posix_bucket_cache.cc:135]: I can't remember;  I think I intended to be testing different invariants

File: src/rgw/driver/posix/rgw_sal_posix.cc [hunk 3]
```diff
@@ -624,52 +630,70 @@ int POSIXBucket::list(const DoutPrefixProvider* dpp, ListParams& params, int max
...
-
-      results.objs.push_back(*e);
+      rgw_bucket_dir_entry bde{};
+      bde.key.name = decode_name(name);
+      bde.ver.pool = 1;
+      bde.ver.epoch = 1;
+      bde.exists = true;
+      bde.meta.category = RGWObjCategory::Main;
+      bde.meta.size = stx.stx_size;
+      bde.meta.mtime = ceph::real_clock::from_time_t(stx.stx_mtime.tv_sec);
```
Comment (mattbenjamin) [inline src/rgw/driver/posix/rgw_sal_posix.cc:649]: if we have a full resolution timespec, why not use it?

Comment (dang) [reply src/rgw/driver/posix/rgw_sal_posix.cc:649]: We could.  statx() doesn't use a normal timestamp, it has it's own struct, so I would have had to open-code it, and I was lazy.

Comment (mattbenjamin) [reply src/rgw/driver/posix/rgw_sal_posix.cc:649]: ah, ok

**Review by dang — APPROVED** [review]

**Comment by dang** [comment]
I'm able to crash this.  What I did was:

Make a bucket (dang) and put some objects in it using s3cmd.  Worked fine.
Added some object so bucket outside of RGW while it was running.  Worked fine.
Stopped RGW, added some objects to bucket, restarted, worked fine.
Stopped RGW, made a new directory (foo) and put a file it it.
Started RGW
s3cmd ls s3://foo/ -> crash
Crash persists if I restart RGW and list the bucket.

Backtrace isn't useful, it's just a thread starting up:

[?2004h[?2004l[?2004h(gdb) bt
#0  0x00007fa1c1a93292 in raise () from /lib64/libc.so.6
#1  0x00007fa1c1a7c8a4 in abort () from /lib64/libc.so.6
#2  0x00007fa1c1e25a46 in __gnu_cxx::__verbose_terminate_handler() [clone .cold] () from /
lib64/libstdc++.so.6
#3  0x00007fa1c1e3129c in __cxxabiv1::__terminate(void (*)()) () from /lib64/libstdc++.so.
6
#4  0x00007fa1c1e302f9 in __cxa_call_terminate () from /lib64/libstdc++.so.6
#5  0x00007fa1c1e30a27 in __gxx_personality_v0 () from /lib64/libstdc++.so.6
#6  0x00007fa1c1c364a4 in _Unwind_RaiseException_Phase2 () from /lib64/libgcc_s.so.1
#7  0x00007fa1c1c36ede in _Unwind_Resume () from /lib64/libgcc_s.so.1
#8  0x00005626b4c4f5d9 in boost::asio::detail::scheduler::run (this=0x5626b7796700, ec=...
) at boost/include/boost/asio/detail/impl/scheduler.ipp:214
#9  0x00005626b4c505ef in boost::asio::io_context::run (this=0x5626b7a71c90) at boost/incl
ude/boost/asio/impl/io_context.ipp:63
#10 0x00005626b4c33e53 in operator() (__closure=0x5626b778eb18) at /home/dang/src/ceph/mai
n/wip-rgw-sal-gpfs/ceph/src/rgw/rgw_asio_frontend.cc:1081
#11 0x00005626b4c49c6e in std::__invoke_impl<void, (anonymous namespace)::AsioFrontend::ru
n()::<lambda()> >(std::__invoke_other, struct {...} &&) (__f=...) at /usr/include/c++/11/b
its/invoke.h:61
#12 0x00005626b4c49c23 in std::__invoke<(anonymous namespace)::AsioFrontend::run()::<lambd
a()> >(struct {...} &&) (__fn=...) at /usr/include/c++/11/bits/invoke.h:96
#13 0x00005626b4c49bd0 in std::thread::_Invoker<std::tuple<(anonymous namespace)::AsioFron
tend::run()::<lambda()> > >::_M_invoke<0>(std::_Index_tuple<0>) (this=0x5626b778eb18) at /
usr/include/c++/11/bits/std_thread.h:253
#14 0x00005626b4c49a4c in std::thread::_Invoker<std::tuple<(anonymous namespace)::AsioFron
tend::run()::<lambda()> > >::operator()(void) (this=0x5626b778eb18) at /usr/include/c++/11
/bits/std_thread.h:260
#15 0x00005626b4c45b20 in std::thread::_State_impl<std::thread::_Invoker<std::tuple<(anony
mous namespace)::AsioFrontend::run()::<lambda()> > > >::_M_run(void) (this=0x5626b778eb10)
 at /usr/include/c++/11/bits/std_thread.h:211
 #16 0x00007fa1c1e5dc84 in execute_native_thread_routine () from /lib64/libstdc++.so.6
 #17 0x00007fa1c20e6299 in start_thread () from /lib64/libpthread.so.0
 #18 0x00007fa1c1b566a3 in clone () from /lib64/libc.so.6
 [?2004h(gdb)

I'll play with this a bit, and see what I can find out.  (Note, this is an ancient machine.  It's Fedora 34 (!), so it's old libraries and compilers.  I was planning on upgrading this machine soon.)

**Comment by dang** [comment]
> Crash persists if I restart RGW and list the bucket.

Oh, and crash persists even if I nuke the database root.

**Comment by dang** [comment]
It appears to be an OOM exception:

Thread 28 "radosgw" hit Catchpoint 1 (exception thrown), 0x00007f7aeb2ae651 in __cxa_throw () from /lib64/libstdc++.so.6
(gdb) bt                                                                                                                                
#0  0x00007f7aeb2ae651 in __cxa_throw () from /lib64/libstdc++.so.6                                                                                                                                                                                                              #1  0x00007f7aeb616222 in (anonymous namespace)::handle_oom(void* (*)(void*), void*, bool, bool) [clone .cold] () from /lib64/libtcmalloc.so.4                                                                                                                                   #2  0x00007f7aeb639c15 in tcmalloc::allocate_full_cpp_throw_oom(unsigned long) ()                                
   from /lib64/libtcmalloc.so.4                                                                                                         
#3  0x000055f255806d0d in std::__new_allocator<char>::allocate (this=0x7f77245c5cc0,
    __n=0x7f77245c6071) at /usr/include/c++/12/bits/new_allocator.h:137
#4  0x000055f2558049bc in std::allocator<char>::allocate (__n=0x7f77245c6071,
    this=0x7f77245c5cc0) at /usr/include/c++/12/bits/allocator.h:188 
#5  std::allocator_traits<std::allocator<char> >::allocate (__a=..., __n=0x7f77245c6071)
    at /usr/include/c++/12/bits/alloc_traits.h:464
#6  0x000055f2558048e4 in std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::_M_create (this=0x7f77245c5cc0,
    __capacity=@0x7f77245c5538: 0x7f77245c6070, __old_capacity=0x0)
    at /usr/include/c++/12/bits/basic_string.tcc:155
#7  0x000055f255805513 in std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::_M_construct<char const*> (this=0x7f77245c5cc0,
    __beg=0x7f77245c60a0 "\240\071\204[\362U",
    __end=0xfeee48b8c110 <error: Cannot access memory at address 0xfeee48b8c110>)
    at /usr/include/c++/12/bits/basic_string.tcc:225
#8  0x000055f25582a395 in std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::basic_string (this=0x7f77245c5cc0,
    __s=0x7f77245c60a0 "\240\071\204[\362U", __n=0x7f77245c6070, __a=...)
    at /usr/include/c++/12/bits/basic_string.h:620
#9  0x000055f2558340dc in std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::basic_string (this=0x7f77245c5cc0, __svw=..., __a=...)
    at /usr/include/c++/12/bits/basic_string.h:188
#10 0x000055f25582a775 in std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::basic_string<std::basic_string_view<char, std::char_traits<char> >, void> (this=0x7f77245c5cc0,
    __t="\240\071\204[\362U\000\000\000\000\000\000\000\000\000\000\265\315'V\362U\000\000J\321'V\362U\000\000\230\272\340X\362U\000\000\000\200\312\371\332\300\251\344\360`\\$w\177\000\000\250s\\$w\177\000\000\240c\\$w\177\000\000\372\246\313U\362U\000\000\020^\020X\362U\
000\000@\213\\$w\177\000\000\020^\020X\362U\000\000@\213\\$w\177\000\000\330_\\$w\177\000\000\000\271\340X\362U\000\000\350_\\$w\177\000\000(\340\222[\362U\000\000(\340\222[\362U\000\000(\340\222[\362U\000\000\340_\\$w\177\000\000@\340\222[\362U\000\000\340_\\$w\177\000\00
0\000\000\000\000\000\000\000\000\350_\\$w\177\000\000"..., __a=...) 
    at /usr/include/c++/12/bits/basic_string.h:787
#11 0x000055f2562819e3 in file::listing::BucketCache<rgw::sal::POSIXDriver, rgw::sal::POSIXBucket>::list_bucket(DoutPrefixProvider const*, rgw::sal::POSIXBucket*, rgw::sal::Bucket::ListParams&, int, fu2::abi_310::detail::function<fu2::abi_310::detail::config<true, false, 1
6ul>, fu2::abi_310::detail::property<true, false, int (rgw_bucket_dir_entry const&) const> >)::{lambda()#1}::operator()() const (__closure=0x7f77245c5f80)
    at /home/dang/src/ceph/main/wip-rgw-sal-gpfs/ceph/src/rgw/driver/posix/bucket_cache.h:405
#12 0x000055f2562841e9 in file::listing::BucketCache<rgw::sal::POSIXDriver, rgw::sal::POSIXBucket>::list_bucket(DoutPrefixProvider const*, rgw::sal::POSIXBucket*, rgw::sal::Bucket::ListParams&, int, fu2::abi_310::detail::function<fu2::abi_310::detail::config<true, false, 1
6ul>, fu2::abi_310::detail::property<true, false, int (rgw_bucket_dir_entry const&) const> >) (this=0x55f25810a3c0, dpp=0x55f258e0b900, sal_bucket=0x55f25b95e000, params=...,
    max=0x3e8, each_func=...)
    at /home/dang/src/ceph/main/wip-rgw-sal-gpfs/ceph/src/rgw/driver/posix/bucket_cache.h:434
#13 0x000055f25626bcb6 in rgw::sal::POSIXBucket::list (this=0x55f25b95e000,
    dpp=0x55f258e0b900, params=..., max=0x3e8, results=..., y=...)
    at /home/dang/src/ceph/main/wip-rgw-sal-gpfs/ceph/src/rgw/driver/posix/rgw_sal_posix.cc:683
#14 0x000055f255cba6fa in RGWListBucket::execute (this=0x55f258e0b900, y=...)
    at /home/dang/src/ceph/main/wip-rgw-sal-gpfs/ceph/src/rgw/rgw_op.cc:3029
#15 0x000055f2559ed4eb in rgw_process_authenticated (handler=0x55f25b958db0,
    op=@0x7f77245c6658: 0x55f258e0b900, req=0x7f77245c75f0, s=0x7f77245c6950, y=...,
    driver=0x55f2580c5400, skip_retarget=0x0)
    at /home/dang/src/ceph/main/wip-rgw-sal-gpfs/ceph/src/rgw/rgw_process.cc:255
#16 0x000055f2559f00a6 in process_request (penv=..., req=0x7f77245c75f0,
    frontend_prefix="", client_io=0x7f77245c76a0, yield=..., scheduler=0x55f2580d2ff8,
    user=0x7f77245c7800, latency=0x7f77245c74a8, http_ret=0x7f77245c749c)
    at /home/dang/src/ceph/main/wip-rgw-sal-gpfs/ceph/src/rgw/rgw_process.cc:392
#17 0x000055f25587525d in (anonymous namespace)::handle_connection<boost::asio::basic_stream_socket<boost::asio::ip::tcp, boost::asio::io_context::basic_executor_type<std::allocator<void>, 0> > > (context=..., env=..., stream=..., timeout=..., header_limit=0x4000,
    buffer=..., is_ssl=0x0, pause_mutex=..., scheduler=0x55f2580d2ff8, uri_prefix="",
--Type <RET> for more, q to quit, c to continue without paging--
    ec=..., yield=...)
    at /home/dang/src/ceph/main/wip-rgw-sal-gpfs/ceph/src/rgw/rgw_asio_frontend.cc:284
#18 0x000055f25586f7d7 in operator() (__closure=0x55f25b914038, yield=...)
    at /home/dang/src/ceph/main/wip-rgw-sal-gpfs/ceph/src/rgw/rgw_asio_frontend.cc:1055
#19 0x000055f25587dc8e in operator() (__closure=0x7f77245c8f18, c=...)
    at /home/dang/src/ceph/main/wip-rgw-sal-gpfs/ceph/src/spawn/include/spawn/impl/spawn.hpp:390
#20 0x000055f255880bcf in std::__invoke_impl<boost::context::continuation, spawn::detail::spawn_helper<boost::asio::executor_binder<void (*)(), boost::asio::strand<boost::asio::io_context::basic_executor_type<std::allocator<void>, 0> > >, (anonymous namespace)::AsioFronten
d::accept(Listener&, boost::system::error_code)::<lambda(yield_context)>, boost::context::basic_protected_fixedsize_stack<boost::context::stack_traits> >::operator()()::<lambda(boost::context::continuation&&)>&, boost::context::continuation>(std::__invoke_other, struct {..
.} &) (__f=...) at /usr/include/c++/12/bits/invoke.h:61
#21 0x000055f255880a82 in std::__invoke<spawn::detail::spawn_helper<boost::asio::executor_binder<void (*)(), boost::asio::strand<boost::asio::io_context::basic_executor_type<std::allocator<void>, 0> > >, (anonymous namespace)::AsioFrontend::accept(Listener&, boost::system:
:error_code)::<lambda(yield_context)>, boost::context::basic_protected_fixedsize_stack<boost::context::stack_traits> >::operator()()::<lambda(boost::context::continuation&&)>&, boost::context::continuation>(struct {...} &) (__fn=...)
    at /usr/include/c++/12/bits/invoke.h:97
#22 0x000055f2558808f7 in std::invoke<spawn::detail::spawn_helper<boost::asio::executor_binder<void (*)(), boost::asio::strand<boost::asio::io_context::basic_executor_type<std::allocator<void>, 0> > >, (anonymous namespace)::AsioFrontend::accept(Listener&, boost::system::e
rror_code)::<lambda(yield_context)>, boost::context::basic_protected_fixedsize_stack<boost::context::stack_traits> >::operator()()::<lambda(boost::context::continuation&&)>&, boost::context::continuation>(struct {...} &) (__fn=...)
    at /usr/include/c++/12/functional:111
#23 0x000055f2558806eb in boost::context::detail::record<boost::context::continuation, boost::context::basic_protected_fixedsize_stack<boost::context::stack_traits>, spawn::detail::spawn_helper<boost::asio::executor_binder<void (*)(), boost::asio::strand<boost::asio::io_co
ntext::basic_executor_type<std::allocator<void>, 0> > >, (anonymous namespace)::AsioFrontend::accept(Listener&, boost::system::error_code)::<lambda(yield_context)>, boost::context::basic_protected_fixedsize_stack<boost::context::stack_traits> >::operator()()::<lambda(boost
::context::continuation&&)> >::run(boost::context::detail::fcontext_t) (
    this=0x7f77245c8f00, fctx=0x7f6faf5f8650)
    at /home/dang/src/ceph/main/wip-rgw-sal-gpfs/build/ceph/boost/include/boost/context/continuation_fcontext.hpp:143
#24 0x000055f2558801dc in boost::context::detail::context_entry<boost::context::detail::record<boost::context::continuation, boost::context::basic_protected_fixedsize_stack<boost::context::stack_traits>, spawn::detail::spawn_helper<boost::asio::executor_binder<void (*)(), 
boost::asio::strand<boost::asio::io_context::basic_executor_type<std::allocator<void>, 0> > >, (anonymous namespace)::AsioFrontend::accept(Listener&, boost::system::error_code)::<lambda(yield_context)>, boost::context::basic_protected_fixedsize_stack<boost::context::stack_
traits> >::operator()()::<lambda(boost::context::continuation&&)> > >(boost::context::detail::transfer_t) (t=...)
    at /home/dang/src/ceph/main/wip-rgw-sal-gpfs/build/ceph/boost/include/boost/context/continuation_fcontext.hpp:80
#25 0x000055f2568adbef in make_fcontext ()
#26 0x0000000000000000 in ?? ()

**Comment by mattbenjamin** [comment]
ok, that should be fixed

**Comment by mattbenjamin** [comment]
@dang I fixed my real build errors, now I'm getting a bogus one (i.e., one on the base branch when I last rebased), or so I think?

**Comment by mattbenjamin** [comment]
jenkins test make check

**Review by dang — APPROVED** [review]

**Comment by dang** [comment]
Not sure why make check is marked failed when it didn't fail...

**Comment by mattbenjamin** [comment]
maybe it's still related to the submodule;  I'm switching that out for https://github.com/Martchus/lmdb-safe, which is currently maintained and has other bugfixes (hopefully, I need to get it to commit a minor build fix).

**Comment by mattbenjamin** [comment]
@dang ^^

**Pull request merged**